Can’t pass a union type as `styled-component`’s type argument

- [x] I tried using the `@types/styled-components` package and had problems.
- [x] I tried using the latest stable version of tsc: `3.6.4`
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see [`Definitions by:` in `types/styled-components/index.d.ts`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/4a4375a3acedf893aaed1ee39fcec7f48b2ec8be/types/styled-components/index.d.ts#L3-L10)) so they can respond.
  - Authors: @Igorbek, @Igmat, @lavoaster, @Jessidhia, @jkillian, @eps1lon, @flavordaaave, @wagerfield

From https://github.com/styled-components/styled-components/issues/2823:

<!-- BUG TEMPLATE -->
## Environment

### System:
 - OS: macOS 10.15
 - CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
 - Memory: 8.08 GB / 32.00 GB
 - Shell: 5.7.1 - /bin/zsh
### Binaries:
 - Node: 12.12.0 - /usr/local/bin/node
 - npm: 6.11.3 - /usr/local/bin/npm
### npm Packages:
 - react: ^16.10.2 => 16.10.2
 - styled-components: ^4.4.0 => 4.4.0
 - @types/react: ^16.9.9 => 16.9.9
 - @types/styled-components: ^4.1.19 => 4.1.19
 - typescript: ^3.6.4 => 3.6.4

## Reproduction

[TypeScript playground](https://www.typescriptlang.org/play/?jsx=2#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wChRJY4BnGATwBskATObXA2xlgWjV0gA7JIJjV8AblKkA9DLgAVOmCTVyopFEzokiFMEF0AIijoAFHGGpwA3qTgO49FQC44AIij7D7qY5oAFhAA7gCqIABGRAwMKACS-IJuERAQTCiCUgC+6jCa2mi6AMoAroKGJuaW1nb+zkhu7tRlPn6O1EHBpYLUaEQiCRBJcClpqJmkOfVwAOqoMAGaM8DMAOZIMBYQVnAAvHoGxqZbOwA+cN0Vx9VSsvIActAgKAxw-OBDImKkmGUYwENZvNFlBlmsNgAKMDVNxzFALJYrdabaoASjcACkigANAB0AFEmCAvrZ7I4mPBgkiNnsPFTwTBfNJ-MBMHAodVcdNdjyPF5Du5UaT-P56cjadDttRcR0QuEokgYvFEnAAPweQBAZIB4P-ccEaJUi0ViTP8WTgiuoulZ7MlVi5yl0PP2TRadEFwpFDjFNP2tulsq6ZV6-UEg0Eao8gD4NwAvu7rGs0en0kCITY4cmSHEQYCUoOGADzMYAANwAfMhiZFNHA6BAc7ZvTAsgBCPMyQul7KkJAAD0o8EStAOVzoAGEBJ9RLS83CEaDqfB6rtPN43YE5YbFbEw7sbDAoCUkGaZCWpD2+28hoPLkdR+PhJP9tPgYiGU4HUuEz414HEyHt7v90POBj1uOQLnoJhWHeIQvjUAd4CKCCWBnEEwXFfYuEgiEUJfZFUQAA3wrte2gftLwQpDmAQFdKjHD573gR9EO4ZgcLnV9F2XAVv3lI1lSGHc9wPI8T2I894PAljr1ou8SSYyi2LQmlOM-VcA26YNk1DRJBMAkSJCAA)

<details>
<summary><b>Source</b>, in case the TypeScript playground link above doesn’t work</summary>

```TypeScript
import React from 'react'
import styled from 'styled-components';

// Types
interface RainyDayProps {
    type: "rainy";
    showUmbrellaIcon: boolean;
}
interface SunnyDayProps {
    type: "sunny";
    showSunscreenIcon: boolean;
}
type WeatherWidgetProps = RainyDayProps | SunnyDayProps;

// Normal components
function WeatherWidget(props: WeatherWidgetProps): JSX.Element {
    let widget = "widget";

    if (props.type === "rainy") {
        widget = props.showUmbrellaIcon ? "☂️" : "umbrella";
    } else if (props.type === "sunny") {
        widget = props.showSunscreenIcon ? "🧴" : "sunscreen";
    }

    return <div>Remember your {widget}!</div>;
}
export const RainyDayComponent = <WeatherWidget type="rainy" showUmbrellaIcon={true} />;
export const SunnyDayComponent = <WeatherWidget type="sunny" showSunscreenIcon={true} />;

// Styled components
const StyledWeatherWidget = styled(WeatherWidget)``
export const StyledRainyDayComponent = <StyledWeatherWidget type="rainy" showUmbrellaIcon={true} />;
export const StyledSunnyDayComponent = <StyledWeatherWidget type="sunny" showSunscreenIcon={true} />;
```
</details>

## Steps to reproduce

When a union type is passed as `styled-component`’s type argument, the TypeScript compiler errors. See the TypeScript playground link above—`StyledWeatherWidget` has red squiggles.

## Expected Behavior

Both interfaces in the union type are valid contracts for the type of the styled React component’s props. As shown in the playground, the unwrapped React component accepts either set of props. Wrapping the React component with `styled-components` should not cause a TypeScript compiler error.

## Actual Behavior

```sh
index.tsx:37:4 - error TS2322: Type '{ type: "rainy"; showUmbrellaIcon: boolean; }' is not assignable to type 'IntrinsicAttributes & Pick<Pick<WeatherWidgetProps, "type"> & Partial<Pick<WeatherWidgetProps, never>>, "type"> & { theme?: any; } & { as?: "symbol" | "object" | "big" | ... 173 more ... | undefined; }'.
  Property 'showUmbrellaIcon' does not exist on type 'IntrinsicAttributes & Pick<Pick<WeatherWidgetProps, "type"> & Partial<Pick<WeatherWidgetProps, never>>, "type"> & { theme?: any; } & { as?: "symbol" | "object" | "big" | ... 173 more ... | undefined; }'.

37   <StyledWeatherWidget type="rainy" showUmbrellaIcon={true} />
      ~~~~~~~~~~~~~~~~~~~

index.tsx:40:4 - error TS2322: Type '{ type: "sunny"; showSunscreenIcon: boolean; }' is not assignable to type 'IntrinsicAttributes & Pick<Pick<WeatherWidgetProps, "type"> & Partial<Pick<WeatherWidgetProps, never>>, "type"> & { theme?: any; } & { as?: "symbol" | "object" | "big" | ... 173 more ... | undefined; }'.
  Property 'showSunscreenIcon' does not exist on type 'IntrinsicAttributes & Pick<Pick<WeatherWidgetProps, "type"> & Partial<Pick<WeatherWidgetProps, never>>, "type"> & { theme?: any; } & { as?: "symbol" | "object" | "big" | ... 173 more ... | undefined; }'.

40   <StyledWeatherWidget type="sunny" showSunscreenIcon={true} />
      ~~~~~~~~~~~~~~~~~~~
```

--- a/types/styled-components/index.d.ts
+++ b/types/styled-components/index.d.ts
@@ -346,7 +346,7 @@ export type ThemedCssFunction<T extends object> = BaseThemedCssFunction<
 >;
 
 // Helper type operators
-type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
+type Omit<T, K extends keyof T> = T extends any ? Pick<T, Exclude<keyof T, K>> : never;
 type WithOptionalTheme<P extends { theme?: T }, T> = Omit<P, "theme"> & {
     theme?: T;
 };

--- a/types/styled-components/test/index.tsx
+++ b/types/styled-components/test/index.tsx
@@ -1052,8 +1052,7 @@ function unionTest() {
         font-size: ${props => props.kind === 'book' ? 16 : 14}
     `;
 
-    // undesired, fix was reverted because of https://github.com/Microsoft/TypeScript/issues/30663
-    <StyledReadable kind="book" author="Hejlsberg" />; // $ExpectError
+    <StyledReadable kind="book" author="Hejlsberg" />;
     <StyledReadable kind="magazine" author="Hejlsberg" />; // $ExpectError
 }
 

